docs/api/specs/webConnectionProtocol.md

```diff
@@ -0,0 +1,382 @@
+  //parent window
+  if (!!window.opener) { candidates.push(window.opener); }
+
+  //parent frames
+  let currentWin = window;
+  while (currentWin.parent !== currentWin){
+    candidates.push(currentWin.parent);
+    currentWin = currentWin.parent;
+  }
+
+  //parent window of top-level parent frame
+  if (window !== currentWin && !!currentWin.opener) { 
+    candidates.push(currentWin.opener); 
+  }
+  return candidates;
+}
+```
+
+Setup a timer for specified timeout, and then for each `candidate` found, attempt to establish communication with it as follows:
+
+  1. Add a listener (`candidate.addEventListener("message", (event) => {})`) to receive response messages from the `candidate`.
+  2. Send a [`"WCP1Hello"`](https://fdc3.finos.org/schemas/next/api/WCP1Hello.schema.json) message to it via `postMessage`.
+
+  ```ts
+  const hello = {
+      type: "WCP1Hello",
+      payload: {
+          identityUrl: identityUrl,
+          actualUrl: actualUrl,
+          fdc3Version: "2.2",
+          intentResolver: true
+          channelSelector: true
+      },
+      meta: {
+          connectionAttemptUuid: "bc96f1db-9b2b-465f-aab3-3870dc07b072"
+          timestamp: "2024-09-09T11:44:39+00:00"
+      }
+  };
+  candidate.postMessage(hello, { targetOrigin: "*" });
+  ```
+
+  Note that the `targetOrigin` is set to `*` as the origin of the Desktop Agent is not known at this point.
+  3. Accept the first correct response received from a candidate. Correct responses MUST correspond to either the [`WCP2LoadUrl`](https://fdc3.finos.org/schemas/next/api/WCP2LoadUrl.schema.json) or [`WCP3Handshake`](https://fdc3.finos.org/schemas/next/api/WCP3Handshake.schema.json) message schemas and MUST quote the same `meta.connectionAttemptUuid` value provided in the original `WCP1Hello` message. Stop the timeout when a correct response is received. If no response is received from any candidate, the `getAgent()` implementation MAY retry sending the `WCP1Hello` message periodically until the timeout is reached.
+  4. If a [`WCP3Handshake`](https://fdc3.finos.org/schemas/next/api/WCP3Handshake.schema.json) was received in the previous step, skip this this step and move on to 5. However, If a [`WCP2LoadUrl`](https://fdc3.finos.org/schemas/next/api/WCP2LoadUrl.schema.json) was received in the previous step:
+      - Create a hidden iframe within the page, set its URL to the URL provided by the `payload.iframeUrl` field of the message and add a handler to run when the iframe has loaded:
+          ```ts
+          const loadIframe = (url, loadedHandler): WindowProxy => {
+            const ifrm = document.createElement("iframe");
+            iframe.onload = loadedHandler;
+            ifrm.src = url;
+            ifrm.style.width = "0";
+            ifrm.style.height = "0";
+            ifrm.style.visibility = "0";
+            ifrm.ariaHidden="true";
+            document.body.appendChild(ifrm);
+            return ifrm.contentWindow;
+          }
+          ```
```

Comment:
- Missing types for arguments.  Given that we're specifying the return type for the function explicitly, it would be better to do do for the url and loadedHandler as well.
- I haven't looked at the code enough to see where this is consumed, but when I create functions which accept handlers as arguments, I always do runtime validation e.g. testing  typeof loadedHandler !== "function"  and throwing a runtime error if true.
- Missing semicolon after closing brace for function definition.
- Missing spaces around equals sign when setting ariaHidden.

Reply:
fixed in c5046e28e12134f2c43ba7261f3df0303d45ea4b
I didn't validate the handler type as the typing should really catch that at build time (not going to be accepting handlers from outside the implementation

Reply:
Ah I see.  Yes, if all handlers are inside the implementation then I agree runtime validation not necessary as the added typing will catch this.